module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,688 @@
+		if attempt > 0 {
+			e.metrics.FetchRetried()
+		}
+		attempt++
+
+		// download execution data for the block
+		err := e.processFetchRequest(ctx, request)
+
+		// don't retry if the blob was invalid
+		if isInvalidBlobError(err) {
+			return err
+		}
+
+		return retry.RetryableError(err)
+	})
+}
+
+func (e *executionDataRequester) processFetchRequest(ctx irrecoverable.SignalerContext, request *jobs.BlockEntry) error {
+	logger := e.log.With().
+		Str("block_id", request.BlockID.String()).
+		Uint64("height", request.Height).
+		Logger()
+	logger.Debug().Msg("processing fetch request")
+
+	result, err := e.results.ByBlockID(request.BlockID)
+
+	// By the time the block is sealed, the ExecutionResult must be in the db
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to lookup execution result for block: %w", err))
+	}
+
+	executionData, err := e.fetchExecutionData(ctx, result.ExecutionDataID, request.Height)
+
+	if isInvalidBlobError(err) {
+		// This means an execution result was sealed with an invalid execution data id (invalid data).
+		// Eventually, verification nodes will verify that the execution data is valid, and not sign the receipt
+		logger.Error().Err(err).
+			Str("execution_data_id", result.ExecutionDataID.String()).
+			Msg("HALTING REQUESTER: invalid execution data found")
+
+		e.metrics.Halted()
+		return err
+	}
+
+	// Some or all of the blob was missing or corrupt. retry
+	if isBlobNotFoundError(err) || errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
+		logger.Error().Err(err).Msg("failed to get execution data for block")
+
+		return err
+	}
+
+	// Any other error is unexpected
+	if err != nil {
+		logger.Error().Err(err).
+			Str("execution_data_id", result.ExecutionDataID.String()).
+			Msg("unexpected error fetching execution data")
+
+		ctx.Throw(err)
+	}
+
+	logger.Debug().Msg("Fetched execution data")
+
+	e.executionDataCache.Add(request.Height, executionData)
+
+	return nil
+}
+
+// fetchExecutionData fetches the ExecutionData by its ID, and times out if fetchTimeout is exceeded
+func (e *executionDataRequester) fetchExecutionData(signalerCtx irrecoverable.SignalerContext, executionDataID flow.Identifier, height uint64) (executionData *state_synchronization.ExecutionData, err error) {
+	ctx, cancel := context.WithTimeout(signalerCtx, e.config.FetchTimeout)
+	defer cancel()
+
+	start := time.Now()
+	e.metrics.ExecutionDataFetchStarted()
+	defer func() {
+		// needs to be run inside a closure so the variables are resolved when the defer is executed
+		e.metrics.ExecutionDataFetchFinished(time.Since(start), err == nil, height)
+	}()
+
+	// Get the data from the network
+	// this is a blocking call, won't be unblocked until either hitting error (including timeout) or
+	// the data is received
+	executionData, err = e.eds.Get(ctx, executionDataID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Write it to the local blobstore
+	_, _, err = e.eds.Add(signalerCtx, executionData)
```

Comment:
I'm a bit confused why `Add` is called right after `Get`. 

If you were able to `Get`, it means `eds` has it, but then why adding it back again if `eds` already has it.

Reply:
This is the way the `ExecutionDataService` works. I was a confused at first too.

`Get` just gets the data from the db, or from the network if it's not already stored. It doesn't actually add data retrieved from the network to the db. That has to be done explicitly with a call to Add.

Reply:
In that case, It think the execution data service could provide Fetch, FetchAndAdd, and Add methods. So here we can use FetchAndAdd, which internally calls Fetch then Add.

Reply:
after discussing with Simon, the Add was unnecessary. I've removed it